csrc/preseg_passes/move_gather.cpp

```diff
@@ -0,0 +1,242 @@
+// clang-format off
+/*
+ * SPDX-FileCopyrightText: Copyright (c) 2024-present NVIDIA CORPORATION & AFFILIATES.
+ * All rights reserved.
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+// clang-format on
+#include <preseg_passes/move_gather.h>
+
+#include <expr_simplifier.h>
+#include <fusion.h>
+#include <ir/builder.h>
+#include <ir/interface_nodes.h>
+#include <ir/internal_base_nodes.h>
+#include <ir/utils.h>
+#include <ops/alias.h>
+#include <ops/arith.h>
+#include <ops/indexing.h>
+#include <ops/utils.h>
+#include <scheduler/utils.h>
+#include <transform_replay.h>
+
+namespace nvfuser::preseg_passes {
+
+namespace {
+
+bool isUnaryPointwiseOp(const Expr* expr) {
+  // Check if the expression is a UnaryOp and a pointwise operation
+  return expr->isA<UnaryOp>() && ir_utils::isPointwiseTvOp(expr);
+}
+
+// For now we allow unary pointwise ops like cast/neg or a squeeze.
+// We only consider squeeze which has a single dim squeezed.
+std::optional<Expr*> getAllowedLookupTvDef(const GatherOp* gather_op) {
+  if (gather_op->lookupTv()->isFusionInput()) {
+    return std::nullopt;
+  }
+
+  auto producing_expr = gather_op->lookupTv()->definition();
+  NVF_ERROR(
+      producing_expr != nullptr, "Def for Lookup tensor view is nullptr.");
+
+  if (isUnaryPointwiseOp(producing_expr)) {
+    return producing_expr;
+  }
+
+  // We handle squeeze op where only one dim is squeezed.
+  if (producing_expr->isA<SqueezeOp>()) {
+    auto squeeze_op = producing_expr->as<SqueezeOp>();
+    if (std::count(
+            squeeze_op->getSqueezeDimFlags().begin(),
+            squeeze_op->getSqueezeDimFlags().end(),
+            true) == 1) {
+      return producing_expr;
+    }
+  }
+
+  return std::nullopt;
+}
+
+auto positionOfSqueezedDim(const std::vector<bool>& squeeze_dim_flags) {
```

Comment:
nit. I don't know if there's any written rule, but I don't like using `auto` for return types of normal functions. It just adds another unknown while reading the code from top to bottom.